11-Backbonejs/06-Adding-Custom-Events.md

```diff
@@ -0,0 +1,259 @@
+1.  When the task Lists' listener 'hears' the event it can delete the card being edited, and put the title & description fields in the input tags.
+
+There will be drawbacks, but we'll see how it works.
+
+### First create an Event listener in TaskView for when the `<div>` in the template is clicked on:
+
+```javascript
+   // in app/views/task_view.js
+  events: {
+    'click .complete-button': 'completeHandler',
+    'click .delete-button': 'deleteTask',
+    'click div': 'editTask'
+  },
+  editTask: function() {
+    // Trigger the editMe event
+    this.trigger('editMe', this);
+  },
+```
+
+Notice that above we are **triggering** an event called `editMe` and passing any listeners a copy of `this` object.  
+
+### Second Listen for the Event & `trigger` a custom Backbone Event when it occurs.
+
+TaskListView needs to listen to the event however, so we'll need to add a Listener to each 'card' in our view and define a callback function.  
+
+```javascript
+    // In app/views/task_list_view.js
+addTask: function(task) {
+    // Create a TaskView object
+  var card = new TaskView({
+    model: task,
+    template: this.taskTemplate
+  });
+  
+    // Add the object to the list of TaskViews  
+  this.cardList.push(card);
+  
+    // Listen for the 'editMe' event.
+  this.listenTo(card, "editMe", this.editCard);
+},
+
+  // EventListener
+editCard: function(card) {
+    // set the title to the selected Card's title
+  this.input.title.val( card.model.get("title"));
+
+
+    // set the description to the selected card's description
+  this.input.description.val(card.model.get("description"));
+
+    // remove the Task from the collection
+  this.model.remove(card.model);
+},
+```
+
+### Third Have `TaskListView` listen for the custom event
+
+Now our TaskListView will listen for any TaskView it contains.  Further when the `editMe` event occurs our `editCard` function will run setting the form fields to the fields in the selected card's model.  Then we remove the card from our collection.  
+
+Why don't we have to remove it from our `cardList` array?  
+
+Further, if we click on cards randomly we end up deleting them...  How could we handle this better?  
+
+
+## Check-in
+
+At this point your views should look like [this:](https://gist.github.com/CheezItMan/14346e3bcf1cb25879341713a849015d)
+
+
+#### Word Of Caution:
+
+Sometimes you may have a View listening for events on a particular Model.  If you remove the View in the course of your program without removing the corresponding Model and without turning off any event listening this will cause the View to hang around in memory as the Model still thinks it has to notify the 'ghost view' about events, even though the View has been removed from the the program.  This is a common form of 'memory leak' in Backbone.  If you use `on()` you will need to make darn sure you stop listening to events manually before removing any listeners.  
+
+A good reason to use `listenTo()`!
```

Comment:
As above, if we're going to strongly suggest that they use `listenTo` instead of `on` so as to avoid the problems mentioned above, which they likely do not understand at this point, maybe we should just not teach them to use `on`.